collector/cache.go

```diff
@@ -0,0 +1,453 @@
+			"(FastReadNotPossiblesPersec)",
+			nil,
+			nil,
+		),
+		FastReadResourceMissesPersec: prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, subsystem, "fast_read_resource_misses_persec"),
+			"(FastReadResourceMissesPersec)",
+			nil,
+			nil,
+		),
+		FastReadsPersec: prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, subsystem, "fast_reads_persec"),
+			"(FastReadsPersec)",
+			nil,
+			nil,
+		),
+		LazyWriteFlushesPersec: prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, subsystem, "lazy_write_flushes_persec"),
+			"(LazyWriteFlushesPersec)",
+			nil,
+			nil,
+		),
+		LazyWritePagesPersec: prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, subsystem, "lazy_write_pages_persec"),
+			"(LazyWritePagesPersec)",
+			nil,
+			nil,
+		),
+		MDLReadHitsPercent: prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, subsystem, "mdl_read_hits_percent"),
+			"(MDLReadHitsPercent)",
+			nil,
+			nil,
+		),
+		MDLReadsPersec: prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, subsystem, "mdl_reads_persec"),
+			"(MDLReadsPersec)",
+			nil,
+			nil,
+		),
+		PinReadHitsPercent: prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, subsystem, "pin_read_hits_percent"),
+			"(PinReadHitsPercent)",
+			nil,
+			nil,
+		),
+		PinReadsPersec: prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, subsystem, "pin_reads_persec"),
+			"(PinReadsPersec)",
+			nil,
+			nil,
+		),
+		ReadAheadsPersec: prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, subsystem, "read_aheads_persec"),
+			"(ReadAheadsPersec)",
+			nil,
+			nil,
+		),
+		SyncCopyReadsPersec: prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, subsystem, "sync_copy_reads_persec"),
+			"(SyncCopyReadsPersec)",
+			nil,
+			nil,
+		),
+		SyncDataMapsPersec: prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, subsystem, "sync_data_maps_persec"),
+			"(SyncDataMapsPersec)",
+			nil,
+			nil,
+		),
+		SyncFastReadsPersec: prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, subsystem, "sync_fast_reads_persec"),
+			"(SyncFastReadsPersec)",
+			nil,
+			nil,
+		),
+		SyncMDLReadsPersec: prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, subsystem, "sync_mdl_reads_persec"),
+			"(SyncMDLReadsPersec)",
+			nil,
+			nil,
+		),
+		SyncPinReadsPersec: prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, subsystem, "sync_pin_reads_persec"),
+			"(SyncPinReadsPersec)",
+			nil,
+			nil,
+		),
+	}, nil
+}
+
+// Collect sends the metric values for each metric
+// to the provided prometheus Metric channel.
+func (c *CacheCollector) Collect(ctx *ScrapeContext, ch chan<- prometheus.Metric) error {
+	if desc, err := c.collect(ch); err != nil {
+		log.Error("failed collecting cache metrics:", desc, err)
+		return err
+	}
+	return nil
+}
+
+// Win32_PerfFormattedData_PerfOS_Cache  docs:
+// - https://docs.microsoft.com/en-us/previous-versions/aa394267(v=vs.85)
+type Win32_PerfFormattedData_PerfOS_Cache struct {
```

Comment:
If we use WMI, we actually want to use the `PerfRawData` classes rather than the formatted ones. 
But it seems we could actually use the new Perflib interface instead of WMI, the counter name is `Cache`. Would you be willing to refactor to use that instead? It is many, many times faster than WMI, and more reliable. You can look at eg the `logical_disk` collector for an example. Unfortunately, there isn't a generator script yet.

Reply:
I'll take a look at refactoring this to Perflib